machine/v1beta1/types_controlplanemachineset.go

```diff
@@ -0,0 +1,244 @@
+// +kubebuilder:printcolumn:name="Updated",type="integer",JSONPath=".status.updatedReplicas",description="Updated Replicas"
+// +kubebuilder:printcolumn:name="Unavailable",type="string",JSONPath=".status.unavailableReplicas",description="Observed number of unavailable replicas"
+// +kubebuilder:printcolumn:name="Age",type="date",JSONPath=".metadata.creationTimestamp",description="Machineset age"
+// Compatibility level 2: Stable within a major release for a minimum of 9 months or 3 minor releases (whichever is longer).
+// +openshift:compatibility-gen:level=2
+type ControlPlaneMachineSet struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   ControlPlaneMachineSetSpec   `json:"spec,omitempty"`
+	Status ControlPlaneMachineSetStatus `json:"status,omitempty"`
+}
+
+type ControlPlaneMachineSetSpec struct {
+	// Replicas defines how many Control Plane Machines should be
+	// created by this ControlPlaneMachineSet.
+	// This field is immutable and cannot be changed after cluster
+	// installation.
+	// The ControlPlaneMachineSet only operates with 3 or 5 node control planes,
+	// 3 and 5 are the only valid values for this field.
+	// +kubebuilder:validation:Enum:=3;5
+	// +kubebuilder:default:=3
+	// +kubebuilder:validation:Required
+	Replicas *int32 `json:"replicas"`
+
+	// Strategy defines how the ControlPlaneMachineSet will update
+	// Machines when it detects a change to the ProviderSpec.
+	// +kubebuilder:default:={type: RollingUpdate}
+	// +optional
+	Strategy ControlPlaneMachineSetStrategy `json:"strategy,omitempty"`
+
+	// FailureDomains is the list of failure domains (sometimes called
+	// availability zones) in which the ControlPlaneMachineSet should balance
+	// the Control Plane Machines.
+	// This will be merged into the ProviderSpec given in the template.
+	// This field is optional on platforms that do not require placement
+	// information, eg OpenStack.
+	// +optional
+	FailureDomains FailureDomains `json:"failureDomains,omitempty"`
+
+	// Label selector for Machines. Existing Machines selected by this
+	// selector will be the ones affected by this ControlPlaneMachineSet.
+	// It must match the template's labels.
+	// This field is considered immutable after creation of the resource.
+	// +kubebuilder:validation:Required
+	Selector metav1.LabelSelector `json:"selector"`
+
+	// Template describes the Control Plane Machines that will be created
+	// by this ControlPlaneMachineSet.
+	// +kubebuilder:validation:Required
+	Template ControlPlaneMachineSetTemplate `json:"template"`
+}
+
+type ControlPlaneMachineSetTemplate struct {
+	// ObjectMeta is the standard object metadata
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// Labels are required to match the ControlPlaneMachineSet selector.
+	// +kubebuilder:validation:Required
+	ObjectMeta metav1.ObjectMeta `json:"metadata"`
```

Comment:
What information is valid to set here and why would I set it?

Reply:
This matches the pattern of the MachineSet, where the labels and annotations in this objectmeta become the labels and annotations on the newly minted machines (much like in a deployment/replicaset/pod relationship).

I'm gonna take a stab and say you're going to suggest we trim this to our own objectmeta object which supports only the fields we need? (We [sort of](https://github.com/openshift/api/blob/354aa98a475c1fcd60b41aaee735da95d7318b42/machine/v1beta1/types_provider.go#L44) did this in MachineSet already, can't recall exactly the history of that, will do some archeology and come back)

Reply:
Looks like we added that because of https://bugzilla.redhat.com/show_bug.cgi?id=1702089, for some reason the CRD generation wasn't generating a schema for the objectmeta back then and so the solution was to embed it in our own types so we could get schema validation. We copied from CAPI so that's why we have the fields we have, even though we only actually use labels and annotations 🤔